### PR TITLE
Added LjCos2 and LJCos non-bonded interaction with test cases

### PR DESCRIPTION
**Pull request by NishchaySuri**
Adresses #951

Description of changes:
 - Changes are made in interactions.pyx and interactions.pxd
 - A class is added to include the lennard-jones cosine2 non-bonded interaction. It is placed in the LENNARD_JONES directive! (Was not sure if this is how you want it structured)
- I have compiled it and is working.

Please tell me if it is okay or if I need to make any changes? I would be glad to contribute
Thanks!
-Nishchay Suri

PR Checklist
------------
 - [ ] Tests?
   - [ ] Interface
   - [ ] Core 
 - [x] Docs?

### CODE REVIEW

**Comment by hmenke** [comment]
Thank you very much for your contribution!  The code looks reasonable.  One of the tests failed because the `LJCOS2` feature is not always enabled, so I added some guards for you.  Please also adhere to the Python [PEP8](https://www.python.org/dev/peps/pep-0008/) naming convention.

Before this can be merged, you have to add a testcase though.
You can get inspired by the old `lj-cos.tcl` testcase.
https://github.com/espressomd/espresso/blob/6cbef0e6b2e102f0fa75c06c6288aea87d620371/testsuite/tcl/lj-cos.tcl#L1-L138

**Comment by NishchaySuri** [comment]
I have added the testcase for Ljcos2 and a data file it pulls from named test_ljcos2.dat
Please check if this is enough or if I need to make any other changes!
Happy to contribute.
Thanks
-Nishchay

**Comment by NishchaySuri** [comment]
I have added the test case for lj-cos2 interaction. It is passing all the tests. Please see if this is all you wanted for the merge!

Happy to contribute!
Thank you
-Nishchay Suri

File: doc/sphinx/inter.rst [hunk 1]
```diff
@@ -233,41 +233,46 @@ on/off during a simulation.
...
 
-.. math:: V(r)=\frac{1}{2}\epsilon\left(cos\left[\alpha(r - r_\mathrm{off})^2 + \beta\right]-1\right),
+:class:`espressomd.interactions.LennardJonesCosInteraction` and
+:class:`espressomd.interactions.LennardJonesCos2Interaction` specifies
+a Lennard-Jones interaction with cosine tail :cite:`soddeman01a`
+between particles of the types *type1* and *type2* . The first variant
+behaves as follows: Until the minimum of the Lennard-Jones potential
+at :math:`r_\mathrm{min} = r_\mathrm{off} + 2^{\frac{1}{6}}\sigma`, it
+behaves identical to the unshifted Lennard-Jones potential
+(:math:`c_\mathrm{shift}=0`). Between and , a cosine is used to
```
Comment (pkreissl) [inline doc/sphinx/inter.rst:252]: Between ? and ?

File: src/python/espressomd/interactions.pyx [hunk 2]
```diff
@@ -313,6 +313,139 @@ IF LENNARD_JONES == 1:
...
+            }
+
+        def type_name(self):
+            return "LennardJonesCos"
+
+        def valid_keys(self):
+            return "epsilon", "sigma", "cutoff", "offset"
+
+        def required_keys(self):
+            return "epsilon", "sigma", "cutoff", "offset"
```
Comment (pkreissl) [inline src/python/espressomd/interactions.pyx]: Is `offset` really required (default is `offset=0`)?

File: src/python/espressomd/interactions.pyx [hunk 3]
```diff
@@ -313,6 +313,139 @@ IF LENNARD_JONES == 1:
...
+                "width"  : 0.}
+                
+        def type_name(self):
+            return "LennardJonesCos2"
+
+        def valid_keys(self):
+            return "epsilon", "sigma", "offset", "width"
+
+        def required_keys(self):
+            return "epsilon", "sigma", "offset", "width"
```
Comment (pkreissl) [inline src/python/espressomd/interactions.pyx]: Again, is explicitly setting `offset` really necessary?

File: src/python/espressomd/interactions.pyx [hunk 4]
```diff
@@ -313,6 +313,139 @@ IF LENNARD_JONES == 1:
...
+        def _get_params_from_es_core(self):
+            cdef ia_parameters * ia_params
+            ia_params = get_ia_param_safe(
+                self._part_types[0],
+                self._part_types[1])
+            return {
+                "epsilon":ia_params.LJCOS_eps,
+                "sigma"  :ia_params.LJCOS_sig,
+                "cutoff" :ia_params.LJCOS_cut,
+                "offset"  :ia_params.LJCOS_offset,
```
Comment (pkreissl) [inline src/python/espressomd/interactions.pyx]: Please use autopep8 you python files.

File: testsuite/test_ljcos2.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
+# This is a unit-test for Lennard Jones Cosine2 non-bonded interaction
```
Comment (pkreissl) [inline testsuite/test_ljcos2.py:1]: Please autopep8 this file!

**Review by pkreissl — CHANGES_REQUESTED** [review]
The python interfaces for both interactions look good. However, I don't think that `offset` needs to be a required key.

Please autopep8 your python files (`autopep8 -i <filename>`).

Concerning the testcase:
Please add your test to the `testsuite/CMakeLists.txt`. Otherwise your test will not be run automatically.

If I am not mistaken, a testcase for the LJCOS interaction is still missing.

File: doc/sphinx/inter.rst (hunk 1 above)
Comment (pkreissl) [reply doc/sphinx/inter.rst:252]: Between :math:`r_\mathrm{min}` and :math:`r_\mathrm{cut}`, ...
(see [ESPResSo User’s Guide](http://espressomd.org/wordpress/wp-content/uploads/2017/06/ESPResSo_3.3.1_UsersGuide.pdf), Section 5.1.4. Lennard-Jones cosine interaction)

**Comment by NishchaySuri** [comment]
All files are autopep8'ed and I also added test for lj cos interaction. 
Also offsets are not 'required' keys anymore
Please see.

Thanks!

**Pull request closed without merging**